### PR TITLE
add cache + sorting + locks to cdc

### PR DESCRIPTION
**Pull request by maha-mstfa**
### Quick checks:

- [ ] There is no other [pull request](https://github.com/conduitio-labs/conduit-connector-dynamodb/pulls) for the same update/change.
- [ ] I have written unit tests.
- [ ] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: iterator/cdc.go [hunk 1]
```diff
@@ -29,224 +32,295 @@ import (
...
+			}
+
+			// sort the buffer after all records were added
+			if changed {
+				sort.Slice(c.cache, func(i, j int) bool {
+					return c.cache[i].Time.Before(c.cache[j].Time)
+				})
+				c.cacheLock.Unlock()
+				c.cacheCond.Signal()
+			}
```
Comment (hariso) [inline iterator/cdc.go]: IIUC, DynamoDB doesn't guarantee that we'll get the records in some order, so we need to order them ourselves?

Comment (hariso) [reply iterator/cdc.go]: Also, the sorting appears to be "local". Let's say we call ` c.streamsClient.GetRecords` and records with timestamps [5, 6, 4]. That's sorted and becomes [4, 5, 6]. Conduit reads a records, so we have [5, 6]. Now we call `c.streamsClient.GetRecords` and get [1, 2, 3]. Everything is sorted together, so we get [1, 2, 3, 5, 6]. Conduit's next record will be 1, which means that Conduit won't get the records ordered anyway.

File: position/position.go [hunk 2]
```diff
@@ -37,7 +37,10 @@ type Position struct {
 	SortKey      string `json:"sort_key"`
 
 	// the record's sequence number in the stream, for the CDC iterator.
-	SequenceNumber string `json:"sequence_number"`
+	SequenceNumberMap map[string]string `json:"sequence_number"`
```
Comment (hariso) [inline position/position.go]: nitpick: would be good that the JSON tag matches the field name a bit more.

File: iterator/cdc.go [hunk 3]
```diff
@@ -29,224 +31,254 @@ import (
...
-	// move to the next shard if the current one is closed
-	if c.shardIndex+1 < len(shards) {
-		c.shardIndex++
-		nextShard := shards[c.shardIndex]
-		// get shard iterator for the new shard
-		return c.getShardIteratorForShard(ctx, *nextShard.ShardId, stypes.ShardIteratorTypeTrimHorizon)
+func (c *CDCIterator) processShard(ctx context.Context, s *shardProcessor) error {
+	defer func() {
+		// clenaup the shard from the SequenceNumberMap when shard is closed.
+		c.cacheLock.Lock()
```
Comment (hariso) [inline iterator/cdc.go]: What's the reason that we use the cache lock here, and not something specific for the last position?

File: iterator/cdc.go [hunk 4]
```diff
@@ -285,35 +316,29 @@ func (c *CDCIterator) getRecMap(item map[string]stypes.AttributeValue) map[strin
...
+		structuredKey[c.sortKey] = image[c.sortKey]
 	}
-	pos := position.Position{
-		IteratorType:   position.TypeCDC,
-		SequenceNumber: *rec.Dynamodb.SequenceNumber,
-		Time:           time.Now(),
-	}
-	cdcPos, err := pos.ToRecordPosition()
+
+	// lock position
```
Comment (hariso) [inline iterator/cdc.go]: We should use a lock here.^

File: iterator/cdc.go [hunk 5]
```diff
@@ -29,224 +31,254 @@ import (
...
+	cache         []recordWithTimestamp
+	cacheLock     sync.Mutex
 	streamArn     string
+	tomb          *tomb.Tomb
+	shards        map[string]*shardProcessor
+	lastPosition  position.Position
+	pollingPeriod time.Duration
+}
+
+type recordWithTimestamp struct {
```
Comment (hariso) [inline iterator/cdc.go:51]: nitpick: it might be good to leave a todo for finding a prettier name and/or renaming the `Time` field (to show what's it about). We can do it in a separate PR. :+1:

File: iterator/cdc.go (hunk 4 above)
Comment (maha-hajja) [reply iterator/cdc.go:335]: do you mean a lock for each write or read on the position?

File: iterator/cdc.go (hunk 3 above)
Comment (maha-hajja) [reply iterator/cdc.go:221]: my bad

**Review by hariso — APPROVED** [review]
:rocket:

**Review by hariso — APPROVED** [review]
